src/coreclr/jit/helperexpansion.cpp

```diff
@@ -0,0 +1,1043 @@
+//    in the cache.
+//
+bool Compiler::fgExpandThreadLocalAccessForCall(BasicBlock* block, Statement* stmt, GenTreeCall* call)
+{
+    assert(call->IsHelperCall());
+    if (!call->IsExpTLSFieldAccess())
+    {
+        return false;
+    }
+
+    CORINFO_THREAD_STATIC_BLOCKS_INFO threadStaticBlocksInfo;
+    info.compCompHnd->getThreadLocalStaticBlocksInfo(&threadStaticBlocksInfo);
+
+    assert(threadStaticBlocksInfo.tlsIndex.accessType == IAT_VALUE);
+    assert(eeGetHelperNum(call->gtCallMethHnd) == CORINFO_HELP_GETSHARED_NONGCTHREADSTATIC_BASE_NOCTOR_OPTIMIZED);
+
+    JITDUMP("Expanding thread static local access for [%06d] in " FMT_BB ":\n", dspTreeID(call), block->bbNum);
+    DISPTREE(call);
+    JITDUMP("\n");
+
+    call->ClearExpTLSFieldAccess();
+    assert(call->gtArgs.CountArgs() == 1);
+
+    // Split block right before the call tree
+    BasicBlock* prevBb       = block;
+    GenTree**   callUse      = nullptr;
+    Statement*  newFirstStmt = nullptr;
+    DebugInfo   debugInfo    = stmt->GetDebugInfo();
+    block                    = fgSplitBlockBeforeTree(block, stmt, call, &newFirstStmt, &callUse);
+    assert(prevBb != nullptr && block != nullptr);
+
+    // Block ops inserted by the split need to be morphed here since we are after morph.
+    // We cannot morph stmt yet as we may modify it further below, and the morphing
+    // could invalidate callUse.
+    while ((newFirstStmt != nullptr) && (newFirstStmt != stmt))
+    {
+        fgMorphStmtBlockOps(block, newFirstStmt);
+        newFirstStmt = newFirstStmt->GetNextStmt();
+    }
+
+    GenTreeLclVar* threadStaticBlockLcl = nullptr;
+
+    // Grab a temp to store result (it's assigned from either fastPathBb or fallbackBb)
+    if (threadStaticBlockLcl == nullptr)
+    {
+        // Define a local for the result
+        unsigned threadStaticBlockLclNum         = lvaGrabTemp(true DEBUGARG("TLS field access"));
+        lvaTable[threadStaticBlockLclNum].lvType = TYP_I_IMPL;
+        threadStaticBlockLcl                     = gtNewLclvNode(threadStaticBlockLclNum, call->TypeGet());
+
+        *callUse = gtClone(threadStaticBlockLcl);
+
+        fgMorphStmtBlockOps(block, stmt);
+        gtUpdateStmtSideEffects(stmt);
+    }
+
+    GenTree* typeThreadStaticBlockIndexValue = call->gtArgs.GetArgByIndex(0)->GetNode();
+
+    void** pIdAddr = nullptr;
+
+    size_t   tlsIndexValue = (size_t)threadStaticBlocksInfo.tlsIndex.addr;
+    GenTree* dllRef        = nullptr;
+
+    if (tlsIndexValue != 0)
+    {
+        dllRef = gtNewIconHandleNode(tlsIndexValue * TARGET_POINTER_SIZE, GTF_ICON_TLS_HDL);
+    }
+
+    // Mark this ICON as a TLS_HDL, codegen will use FS:[cns] or GS:[cns]
+    GenTree* tlsRef = gtNewIconHandleNode(threadStaticBlocksInfo.offsetOfThreadLocalStoragePointer, GTF_ICON_TLS_HDL);
+
+    tlsRef = gtNewIndir(TYP_I_IMPL, tlsRef, GTF_IND_NONFAULTING | GTF_IND_INVARIANT);
```

Comment:
I don't fully understand the final IR shape but I'd recommend to check `GTF_IND_INVARIANT` GT_IND nodes on are they actually invariant or their value depends on initialization state etc, so that JIT in theory can't hoist them before initialization or do some other evil things

Reply:
I recall seeing changes somewhere from @SingleAccretion checking `GTF_IND_INVARIANT` during the dependence analysis we do in lowering, so it would be good to ensure that it is accurate.

Reply:
Was it any of his open PR, because I searched through the code base and just found `GTF_IND_INVARIANT ` being used in value numbering. However, I will double check this on my part.

Reply:
Removed `GTF_IND_INVARIANT` because it will produce `nullptr` first time and valid value 2nd time onwards and hence shouldn't it be marked as invariant.